### PR TITLE
cmd/juju/storage: add binding/life to details

### PR DESCRIPTION
**Pull request by axw**
## Description of change

When displaying the details of storage in YAML or
JSON format, include the binding and life of the entity.
The binding will be simplified to the kind of the entity,
which is "model", "machine", "application", or "unit". 

## QA steps

1. juju bootstrap
2. juju deploy postgresql --storage pgdata=1G
3. juju storage --format=yaml
check that the storage includes life and binding fields

## Documentation changes

This is primarily for debugging, does not need documentation at present.

## Bug reference

None.

### CODE REVIEW

File: apiserver/params/storage.go [hunk 1]
```diff
@@ -544,21 +554,42 @@ type VolumeDetails struct {
 	// VolumeTag is the tag for the volume.
 	VolumeTag string `json:"volume-tag"`
 
+	// BindingTag is the tag of an entity to which the
+	// volume's lifetime is bound.
```
Comment (wallyworld) [inline apiserver/params/storage.go]: include comment about old api version?
here and below

File: apiserver/params/storage.go [hunk 2]
```diff
@@ -544,21 +554,42 @@ type VolumeDetails struct {
...
 	// Storage contains details about the storage instance
 	// that the volume is assigned to, if any.
 	Storage *StorageDetails `json:"storage,omitempty"`
 }
 
+// VolumeAttachmentDetails describes a volume attachment.
+type VolumeAttachmentDetails struct {
+	// NOTE(axw) for backwards-compatibility, this must not be given a
+	// json tag. This ensures that we collapse VolumeAttachmentInfo.
+	VolumeAttachmentInfo
```
Comment (wallyworld) [inline apiserver/params/storage.go]: can we add a todo here to remove this compatibility struct when we get a chance to rev up the api
here and below

File: apiserver/storage/storage.go [hunk 3]
```diff
@@ -190,6 +191,7 @@ func createStorageDetails(st storageAccess, si state.StorageInstance) (*params.S
 				a.Unit().String(),
 				machineTag.String(),
 				location,
+				params.Life(a.Life().String()),
```
Comment (wallyworld) [inline apiserver/storage/storage.go]: would prefer named fields
we use named fields below in StorageDetails struct initialisation

File: cmd/juju/storage/volumelist_test.go [hunk 4]
```diff
@@ -178,7 +178,8 @@ func (s mockListAPI) ListVolumes(machines []string) ([]params.VolumeDetailsListR
 		// storage db-dir/1001, which is attached to unit
 		// abc/0.
 		{
-			VolumeTag: "volume-0-0",
+			VolumeTag:  "volume-0-0",
+			BindingTag: "storage-db-dir-1001",
```
Comment (wallyworld) [inline cmd/juju/storage/volumelist_test.go]: can we include setting the Life attribute

**Comment by axw** [comment]
This PR will be superseded. I'll address comments in the new PR.

**Pull request closed without merging**